File: go-client/internal/protocol/files.go
package protocol

import (
	"crypto/ed25519"
	"encoding/json"
	"fmt"
	"sort"

	"github.com/cisc468/p2p-project/internal/crypto"
	"github.com/cisc468/p2p-project/internal/storage"
)

// HandleListRequest returns a LIST_RESPONSE with all manifests from the file store.
func HandleListRequest(msg map[string]interface{}, fs *storage.FileStore, seq *int) (map[string]interface{}, error) {
	manifests, err := fs.ListManifests()
	if err != nil {
		return nil, err
	}
	files := make([]map[string]interface{}, 0, len(manifests))
	for _, m := range manifests {
		// Convert Manifest struct to map
		data, _ := json.Marshal(m)
		var mm map[string]interface{}
		json.Unmarshal(data, &mm)
		files = append(files, mm)
	}
	*seq++
	return ListResponse(*seq, files), nil
}

// BuildFileChunks creates TRANSFER_CHUNK and TRANSFER_COMPLETE messages for a file.
func BuildFileChunks(fs *storage.FileStore, fileID string, seqStart int) ([]map[string]interface{}, error) {
	fileData, err := fs.LoadFile(fileID)
	if err != nil {
		return nil, err
	}
	manifest, err := fs.LoadManifest(fileID)
	if err != nil {
		return nil, err
	}

	totalChunks := (len(fileData) + ChunkSize - 1) / ChunkSize
	if totalChunks == 0 {
		totalChunks = 1
	}

	var msgs []map[string]interface{}
	seq := seqStart

	for i := 0; i < totalChunks; i++ {
		start := i * ChunkSize
		end := start + ChunkSize
		if end > len(fileData) {
			end = len(fileData)
		}
		chunk := fileData[start:end]
		seq++
		msgs = append(msgs, TransferChunk(seq, fileID, i, totalChunks, chunk))
	}

	seq++
	msgs = append(msgs, TransferComplete(seq, fileID, manifest.SHA256))
	return msgs, nil
}

// ReassembleFile reassembles file data from TRANSFER_CHUNK messages.
func ReassembleFile(chunks []map[string]interface{}) ([]byte, error) {
	// Sort by chunk_index
	sort.Slice(chunks, func(i, j int) bool {
		ci, _ := chunks[i]["chunk_index"].(float64)
		cj, _ := chunks[j]["chunk_index"].(float64)
		return ci < cj
	})

	var data []byte
	for _, c := range chunks {
		chunkDataStr, _ := c["chunk_data"].(string)
		chunkData, err := crypto.B64URLDecode(chunkDataStr)
		if err != nil {
			return nil, fmt.Errorf("decode chunk data: %w", err)
		}
		data = append(data, chunkData...)
	}
	return data, nil
}

// VerifyReceivedFile verifies a received file's integrity and owner signature.
func VerifyReceivedFile(fileData []byte, manifest *storage.Manifest, ownerPKBytes []byte) error {
	if err := storage.VerifyFileIntegrity(fileData, manifest.SHA256); err != nil {
		return err
	}
	ownerPK, err := crypto.BytesToPublicKey(ownerPKBytes)
	if err != nil {
		return err
	}
	return storage.VerifyManifest(manifest, ed25519.PublicKey(ownerPK))
}

File: python-client/p2pshare/protocol/files.py
"""File listing and transfer protocol logic."""

import math

from p2pshare.crypto.keys import b64url_encode, b64url_decode, bytes_to_public_key
from p2pshare.storage.file_store import (
    FileStore,
    verify_manifest,
    verify_file_integrity,
)
from p2pshare.protocol.messages import (
    list_response,
    transfer_response,
    transfer_chunk,
    transfer_complete,
    CHUNK_SIZE,
)
from p2pshare.errors import IntegrityError


def handle_list_request(msg: dict, file_store: FileStore, seq_counter: list) -> dict:
    """Handle a LIST_REQUEST — returns a LIST_RESPONSE with all manifests."""
    manifests = file_store.list_manifests()
    seq_counter[0] += 1
    return list_response(seq_counter[0], manifests)


def build_file_chunks(file_store: FileStore, file_id: str, seq_start: int) -> list[dict]:
    """Build TRANSFER_CHUNK and TRANSFER_COMPLETE messages for a file."""
    file_data = file_store.load_file(file_id)
    manifest = file_store.load_manifest(file_id)

    total_chunks = max(1, math.ceil(len(file_data) / CHUNK_SIZE))
    msgs = []
    seq = seq_start

    for i in range(total_chunks):
        start = i * CHUNK_SIZE
        end = min(start + CHUNK_SIZE, len(file_data))
        chunk = file_data[start:end]
        seq += 1
        msgs.append(transfer_chunk(seq, file_id, i, total_chunks, chunk))

    seq += 1
    msgs.append(transfer_complete(seq, file_id, manifest["sha256"]))
    return msgs


def reassemble_file(chunks: list[dict]) -> bytes:
    """Reassemble file data from TRANSFER_CHUNK messages (already ordered by chunk_index)."""
    chunks_sorted = sorted(chunks, key=lambda c: c["chunk_index"])
    parts = []
    for c in chunks_sorted:
        parts.append(b64url_decode(c["chunk_data"]))
    return b"".join(parts)


def verify_received_file(file_data: bytes, manifest: dict, owner_pk_bytes: bytes) -> None:
    """Verify a received file's integrity and owner signature.

    Raises IntegrityError if either check fails.
    """
    # Verify file hash
    verify_file_integrity(file_data, manifest["sha256"])

    # Verify owner signature
    owner_pk = bytes_to_public_key(owner_pk_bytes)
    verify_manifest(manifest, owner_pk)
